scionlab/forms/attachment_link_form.py

```diff
@@ -0,0 +1,335 @@
+            forwarded_port = (form.cleaned_data['public_port'], False)  # (port, is_bind_port)
+            # Check (public_ip, public_port) clashes
+            if public_ports_map.add(form.cleaned_data['public_ip'],
+                                    form.cleaned_data['public_port']):
+                public_port_clashing_forms.append(form)
+            # Check (bind_ip, bind_port) clashes
+            if form.cleaned_data['bind_port']:
+                forwarded_port = (form.cleaned_data['bind_port'], True)
+                if bind_ports_map.add(form.cleaned_data['bind_ip'],
+                                      form.cleaned_data['bind_port']):
+                    bind_port_clashing_forms.append(form)
+            # Check forwarded_ports clashes
+            if installation_type == UserAS.VM:
+                if forwarded_port[0] in forwarded_ports:
+                    forwarded_port_clashing_forms.append((form, forwarded_port[1]))
+                else:
+                    forwarded_ports.add(forwarded_port[0])
+        if len(isd_set) > 1:
+            raise ValidationError("All attachment points must belong to the "
+                                  "same ISD")
+        elif len(isd_set) < 1:
+            # This means no attachment point has been selected
+            raise ValidationError("Select at least an attachment point")
+        else:
+            self.isd = list(isd_set)[0]
+        # Add public ports clash errors to forms
+        for form in public_port_clashing_forms:
+            form.add_error('public_port',
+                           ValidationError('The port is already in use',
+                                           code='public_port_clash'))
+        # Add bind ports clash errors to forms
+        for form in bind_port_clashing_forms:
+            form.add_error('bind_port',
+                           ValidationError('The port is already in use',
+                                           code='bind_port_clash'))
+        # Add VM forwarded ports clash errors
+        for (form, port_type) in forwarded_port_clashing_forms:
+            port_type_name = 'bind' if port_type else 'public'
+            form.add_error('{}_port'.format(port_type_name),
+                           ValidationError('This port clashes in the VM setup',
+                                           code='forwarded_port_clash'))
+
+    def save(self, user_as, commit=True):
+        self.user_as = user_as
+        for ap_conf in super().save(commit=False):
+            AttachmentLinkForm._create_or_update(user_as, ap_conf)
+        # We save the deleted AP in a set so we only call the cleaning/updating methods once per AP
+        deleted_aps_set = set()
+        for attachment_link in self.deleted_objects:
+            deleted_aps_set.add(attachment_link.interfaceA.AS.attachment_point_info)
+            attachment_link.delete()
+        for ap in deleted_aps_set:
+            ap.split_border_routers()
+            ap.trigger_deployment()
```

Comment:
So at this point, we have a list of the `AttachmentPointConfig` objects that we want, and a bunch of links that should be deleted is that correct? 
I would prefer to move the rest of the logic the model. Here, we'd just call e.g.  `user_as.update_attachments(ap_configs, deleted_links)` and nothing else; this `_create_or_update` function could also move to the model.